core/arch/arm/include/mm/generic_ram_layout.h

```diff
@@ -0,0 +1,196 @@
+ * TEE RAM layout with CFG_WITH_PAGER=y and undefined CFG_TZSRAM_START/_SIZE
+ *
+ *  +----------------------------------+ <-- CFG_TZDRAM_START
+ *  | TEE core secure RAM (TEE_RAM)    |   | | CFG_CORE_TZSRAM_EMUL_SIZE
+ *  +----------------------------------+ --|-'
+ *  |   reserved (for kasan)           |   | TEE_RAM_VA_SIZE
+ *  +----------------------------------+ --'
+ *  | TA RAM / Pagestore (TA_RAM)      |
+ *  +----------------------------------+ <---- align with CORE_MMU_PGDIR_SIZE
+ *  +----------------------------------+ <--
+ *  | SDP test memory (optional)       |   | CFG_TEE_SDP_MEM_SIZE
+ *  +----------------------------------+ <-+ CFG_TZDRAM_START + CFG_TZDRAM_SIZE
+ *
+ *  +----------------------------------+ <-- CFG_SHMEM_START
+ *  | Non-secure static SHM            |   |
+ *  +----------------------------------+   v CFG_SHMEM_SIZE
+ *
+ * ----------------------------------------------------------------------------
+ * TEE RAM layout with CFG_WITH_PAGER=y and define CFG_TZSRAM_START/_SIZE
+ *
+ *  +----------------------------------+ <-- CFG_TZSRAM_START
+ *  | TEE core secure RAM (TEE_RAM)    |   | CFG_TZSRAM_SIZE
+ *  +----------------------------------+ --'
+ *
+ *  +----------------------------------+  <- CFG_TZDRAM_START
+ *  | TA RAM / Pagestore (TA_RAM)      |
+ *  |----------------------------------+ <---- align with CORE_MMU_PGDIR_SIZE
+ *  |----------------------------------+ <--
+ *  | SDP test memory (optional)       |   | CFG_TEE_SDP_MEM_SIZE
+ *  +----------------------------------+ <-+ CFG_TZDRAM_START + CFG_TZDRAM_SIZE
+ *
+ *  +----------------------------------+ <-- CFG_SHMEM_START
+ *  | Non-secure static SHM            |   |
+ *  +----------------------------------+   v CFG_SHMEM_SIZE
+ */
+
+#ifdef CFG_TEE_LOAD_ADDR
+#define TEE_LOAD_ADDR		CFG_TEE_LOAD_ADDR
+#else
+/* Platform specific platform_config.h may set TEE_LOAD_ADDR */
+#endif
+
+#ifdef CFG_TEE_RAM_VA_SIZE
+#define TEE_RAM_VA_SIZE		CFG_TEE_RAM_VA_SIZE
+#else
+#define TEE_RAM_VA_SIZE		CORE_MMU_PGDIR_SIZE
+#endif
+
+#ifdef CFG_SHMEM_SIZE
+#define TEE_SHMEM_SIZE		CFG_SHMEM_SIZE
+#endif
+
+#ifdef CFG_SHMEM_START
+#define TEE_SHMEM_START		CFG_SHMEM_START
+#ifndef CFG_SHMEM_SIZE
+#error CFG_SHMEM_START mandates CFG_SHMEM_SIZE
+#endif
+#endif
+
+#if defined(CFG_TZSRAM_START)
+#define TZSRAM_BASE		CFG_TZSRAM_START
+#define TZSRAM_SIZE		CFG_TZSRAM_SIZE
+#endif
+
+#ifdef CFG_TZDRAM_START
+#if !defined(CFG_WITH_PAGER) || defined(CFG_TZSRAM_START)
+#define TZDRAM_BASE		CFG_TZDRAM_START
+#define TZDRAM_SIZE		CFG_TZDRAM_SIZE
+#else
+#define TZSRAM_BASE		CFG_TZDRAM_START
+#define TZSRAM_SIZE		CFG_CORE_TZSRAM_EMUL_SIZE
+#define TZDRAM_BASE		ROUNDUP(TZSRAM_BASE + TZSRAM_SIZE, \
+					TEE_RAM_VA_SIZE)
+#define TZDRAM_SIZE		(CFG_TZDRAM_START + (CFG_TZDRAM_SIZE - \
+					TZDRAM_BASE))
+#endif
+
+#ifdef CFG_WITH_PAGER
+#define TEE_RAM_START		TZSRAM_BASE
+#define TEE_RAM_PH_SIZE		TZSRAM_SIZE
+#define TA_RAM_START		ROUNDUP(TZDRAM_BASE, CORE_MMU_PGDIR_SIZE)
```

Comment:
I haven't done any calculations here with true values, but have we ever checked that the `ROUNDUP` here and the `ROUNDDOWN` on the next line actually adds up nicely? Spontaneously it feels like values could end up with something that might not match.

Reply:
Actually they do: https://github.com/OP-TEE/optee_os/blob/master/lib/libutils/ext/include/util.h#L55 handles asm and C. I have checked the values computed and looked ok to me. I will test again with non pgdir aligned addresses.

Reply:
I tested various configuration and behaves as expected. TEE_RAM need only 4kB page aligment and TA_RAM mandates MMU pgdir alignment ([`teecore_init_ta_ram()`](https://github.com/OP-TEE/optee_os/blob/3.1.0/core/arch/arm/mm/tee_mmu.c#L918)). This is a good thing cause strange layout (not MByte aligned) will waste a bit but we do fear to consume mapping tables for such fine grained mapping.